runtime/frequency/src/xcm/barrier.rs

```diff
@@ -0,0 +1,38 @@
+use frame_support::traits::{ConstU32, Contains, Everything};
+
+use staging_xcm::latest::prelude::*;
+use staging_xcm_builder::{
+	AllowExplicitUnpaidExecutionFrom, AllowTopLevelPaidExecutionFrom, DenyRecursively,
+	DenyReserveTransferToRelayChain, DenyThenTry, TakeWeightCredit, TrailingSetTopicAsId,
+	WithComputedOrigin,
+};
+
+use crate::xcm::parameters::UniversalLocation;
+
+pub struct ParentOrParentsExecutivePlurality;
+impl Contains<Location> for ParentOrParentsExecutivePlurality {
+	fn contains(location: &Location) -> bool {
+		matches!(location.unpack(), (1, []) | (1, [Plurality { id: BodyId::Executive, .. }]))
+	}
+}
+
+pub type Barrier = TrailingSetTopicAsId<
+	DenyThenTry<
+		DenyRecursively<DenyReserveTransferToRelayChain>,
+		(
+			TakeWeightCredit,
+			WithComputedOrigin<
+				(
+					// we constraint this one so that we limit who can execute instructions
+					AllowTopLevelPaidExecutionFrom<Everything>,
+					AllowExplicitUnpaidExecutionFrom<ParentOrParentsExecutivePlurality>,
+					// ^^^ Parent and its exec plurality get free execution
+					// Subscriptions for version tracking are OK.
+					// AllowSubscriptionsFrom<ParentRelayOrSiblingParachains>,
+				),
+				UniversalLocation,
+				ConstU32<8>,
+			>,
+		),
+	>,
```

Comment:
nit: a unit test for testing barrier would help not not important if other tests cover it

Reply:
Added some placeholder tests for barrier I felt we might have to experiment with it more to make sure we allow only what we seek..

https://github.com/frequency-chain/frequency/pull/2522/files

I can add some comments around tests but thought it would help